File: src/clm/model_evaluation/R/inner-prep-nn-Tc-PubChem.R
setwd("~/git/deepmet")
options(stringsAsFactors = FALSE)
library(argparse)

# dynamically parse arguments
parser = ArgumentParser(prog = 'inner-prep-nn-Tc-PubChem.R')
grid = read.delim("sh/grids/CV/prep-nn-Tc-PubChem.txt")
for (param_name in colnames(grid))
  parser$add_argument(paste0('--', param_name),
                      type = typeof(grid[[param_name]]))
args = parser$parse_args()
print(args)

library(tidyverse)
library(magrittr)
library(data.table)

# read the tabulated data
dat = fread(args$sample_file, data.table = FALSE)

# sample a subset of 500,000 molecules, weighted by frequency
n_tries = 500e3
idxs = sample(seq_len(nrow(dat)), size = n_tries, replace = TRUE, 
              prob = dat$size)
sample = dat[idxs, ]

# read PubChem
pubchem = fread(args$pubchem_file, data.table = FALSE, header = FALSE,
                col.names = c('smiles', 'mass', 'formula'))

# filter PubChem to relevant formulas
fmlas = unique(sample$formula)
pubchem %<>% filter(formula %in% fmlas)

# split by formula
split = pubchem %>% split(.$formula)

# for each molecule, try to find a match from PubChem
matches = pmap_dfr(sample, function(...) {
  current = tibble(...)
  cands = split[[current$formula]]
  if (!current$formula %in% names(split) || nrow(cands) == 0) {
    return(current %>% mutate(source = 'DeepMet'))
  } else {
    match = sample_n(cands, 1)
    ## don't re-sample this molecule again
    # cands = anti_join(cands, match)
    # split[[current$formula]] <<- cands
    ## ^ ignored - sample with replacement
  }
  bind_rows(current %>% mutate(source = 'DeepMet'),
            match %>% mutate(source = 'PubChem'))
})

# write
output_dir = dirname(args$output_file)
if (!dir.exists(output_dir))
  dir.create(output_dir, recursive = TRUE)
write.csv(matches, args$output_file, row.names = FALSE)

File: scripts/model_evaluation/inner_prep_nn_tc_PubChem.py
import argparse
import os
import pandas as pd

parser = argparse.ArgumentParser(description=__doc__)

parser.add_argument("--sample_file", type=str, help="Path to the sampled file")
parser.add_argument("--sample_no", type=int, help="Number of samples to select")
parser.add_argument("--pubchem_file", type=str, help="Path to the PubChem file")
parser.add_argument("--output_file", type=str, help="Path to the save the output file")

args = parser.parse_args()
sample_file = args.sample_file
sample_no = args.sample_no
pubchem_file = args.pubchem_file
output_file = args.output_file


def function(df, split_data):
    current = pd.DataFrame([df])
    current_formula = current['formula'].values[0]
    if ((not current_formula in split_data.keys()) or
        (cands := split_data[current_formula]).shape[0] == 0):
        return current.assign(source="DeepMet")
    match = cands.sample(n=1)
    return pd.concat([current.assign(source="DeepMet"),
               match.assign(source="PubChem")])


sample_file = pd.read_csv(sample_file, delimiter=",")
sample = sample_file.sample(n=sample_no, replace=True, weights=sample_file["size"], ignore_index=True)
pubchem = pd.read_csv(
    pubchem_file, delimiter="\t", header=None, names=["smiles", "mass", "formula"]
)

formulas = set(sample.formula)
pubchem = pubchem[pubchem["formula"].isin(formulas)]
split_data = {formula: df for formula, df in pubchem.groupby('formula')}

result = sample.apply(lambda x: function(x, split_data), axis=1)
matches = pd.concat(result.to_list())
os.makedirs(os.path.dirname(output_file), exist_ok=True)
matches.to_csv(
    output_file,
    index=False
)
